### PR TITLE
842 - Add placeholder generate_computed_files command

### PR DESCRIPTION
**Pull request by avrohomgottlieb**
## Issue Number

https://github.com/AlexsLemonade/scpca-portal/issues/842

## Purpose/Implementation Notes

Added a placeholder for the `generate_computed_files` command. Currently, the command take the functionality of the `load_data` command, as it pertains to the creation of a project's computed files.

This command will serve as the entry point for Batch instances, generating all of a passed project's computed files (according to a predetermined collection of download configs), and uploading them to s3.

## Types of changes

What types of changes does your code introduce?

- New feature (non-breaking change which adds functionality)

## Functional tests

N/A

## Checklist

- [x] Lint and unit tests pass locally with my changes

## Screenshots

N/A

### CODE REVIEW

File: api/scpca_portal/management/commands/generate_computed_files.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+                if update_s3:
+                    s3.upload_output_file(computed_file.s3_key, computed_file.s3_bucket)
+                if clean_up_output_data:
+                    computed_file.clean_up_local_computed_file()
+            computed_file.save()
+
+        # Close DB connection for each thread.
+        connection.close()
+
+    def generate_computed_files(self, **kwargs) -> None:
```
Comment (davidsmejia) [inline api/scpca_portal/management/commands/generate_computed_files.py:96]: Same feedback as before where we want to have the arguments in the function definition.

File: api/scpca_portal/management/commands/generate_computed_files.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        common.INPUT_DATA_PATH.mkdir(exist_ok=True, parents=True)
+
+        # Prepare data output directory.
+        shutil.rmtree(common.OUTPUT_DATA_PATH, ignore_errors=True)
+        common.OUTPUT_DATA_PATH.mkdir(exist_ok=True, parents=True)
+
+        # Prep callback function
+        create_computed_file_kwargs = utils.filter_dict_by_keys(
+            kwargs, {"update_s3", "clean_up_output_data"}
+        )
```
Comment (davidsmejia) [inline api/scpca_portal/management/commands/generate_computed_files.py:110]: Let's remove this and just pass in the defined arguments.

File: api/scpca_portal/management/commands/generate_computed_files.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+            default=settings.PRODUCTION,
+        )
+        parser.add_argument("--max-workers", type=int, default=10)
+        parser.add_argument("--scpca-project-id", type=str)
+        parser.add_argument(
+            "--update-s3", action=BooleanOptionalAction, type=bool, default=settings.UPDATE_S3_DATA
+        )
+
+    def handle(self, *args, **kwargs):
+        self.generate_computed_files(**kwargs)
```
Comment (davidsmejia) [inline api/scpca_portal/management/commands/generate_computed_files.py:68]: Can we break up kwargs here so we aren't passing everything here?

File: api/scpca_portal/management/commands/generate_computed_files.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+    @staticmethod
+    def clean_up_input_data() -> None:
+        shutil.rmtree(common.INPUT_DATA_PATH, ignore_errors=True)
+
+    @staticmethod
+    def clean_up_output_data() -> None:
+        shutil.rmtree(common.OUTPUT_DATA_PATH, ignore_errors=True)
```
Comment (avrohomgottlieb) [inline api/scpca_portal/management/commands/generate_computed_files.py:76]: Note that I updated these functions from their implementations in the `load_data` command.
Over there, they are implemented as follows:
```python
    @staticmethod
    def clean_up_input_data(project) -> None:
        shutil.rmtree(common.INPUT_DATA_PATH / project.scpca_id, ignore_errors=True)

    @staticmethod
    def clean_up_output_data() -> None:
        for path in Path(common.OUTPUT_DATA_PATH).glob("*"):
            path.unlink(missing_ok=True)
```

For code running on Batch instances, it made more sense to me to delete all input and output data directories (and their contents) upon completion of file computation. Thoughts?

File: api/scpca_portal/management/commands/generate_computed_files.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        Uploads file to s3 and cleans up output data depending on passed options.
+        """
+        if computed_file := future.result():
+
+            # Only upload and clean up projects and the last sample if multiplexed
+            if computed_file.project or computed_file.sample.is_last_multiplexed_sample:
+                if update_s3:
+                    s3.upload_output_file(computed_file.s3_key, computed_file.s3_bucket)
+                if clean_up_output_data:
+                    computed_file.clean_up_local_computed_file()
```
Comment (avrohomgottlieb) [inline api/scpca_portal/management/commands/generate_computed_files.py:90]: We use the `clean_up_output_data` kwarg in two different ways:
1. To pass to `create_computed_file` which deletes the recently created computed_file by calling `ComputedFile::clean_up_local_computed_file` after the file has been uploaded to s3
2. To delete all output files at the end of the execution

In an earlier implementation, when we were generated intermediate metadata and readme files, a use case existed to clean up computed files and to clean up "the rest" differently, but now that we don't generated these intermediate files, we should probably remove one of them.

File: api/scpca_portal/management/commands/generate_computed_files.py [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+
+        project.update_downloadable_sample_count()
+
+        if clean_up_input_data:
+            logger.info("Cleaning up input data")
+            self.clean_up_input_data()
+
+        if clean_up_output_data:
+            logger.info("Cleaning up output directory")
+            self.clean_up_output_data()
```
Comment (avrohomgottlieb) [inline api/scpca_portal/management/commands/generate_computed_files.py:154]: Great rid of this guy, it's covered above

File: api/scpca_portal/management/commands/generate_computed_files.py [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+                        config,
+                        sample.get_download_config_file_output_name(config),
+                        sample_lock,
+                    ).add_done_callback(on_get_file)
+
+        project.update_downloadable_sample_count()
+
+        if clean_up_input_data:
+            logger.info("Cleaning up input data")
+            self.clean_up_input_data()
```
Comment (avrohomgottlieb) [inline api/scpca_portal/management/commands/generate_computed_files.py:150]: Move this inside the loop on a per project basis

Comment (avrohomgottlieb) [reply api/scpca_portal/management/commands/generate_computed_files.py:150]: This is not relevant here as we do not loop over projects in this command.

**Review by davidsmejia — APPROVED** [review]
LGTM

**Pull request merged**